plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePartitionedBucketFunction.java

```diff
@@ -0,0 +1,101 @@
+
+import java.lang.invoke.MethodHandle;
+import java.util.List;
+import java.util.stream.IntStream;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Throwables.throwIfUnchecked;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.spi.function.InvocationConvention.InvocationArgumentConvention.BLOCK_POSITION;
+import static io.trino.spi.function.InvocationConvention.InvocationReturnConvention.FAIL_ON_NULL;
+import static io.trino.spi.function.InvocationConvention.simpleConvention;
+import static java.util.Objects.requireNonNull;
+
+public class HivePartitionedBucketFunction
+        implements BucketFunction
+{
+    private final BucketingVersion bucketingVersion;
+    private final int hiveBucketCount;
+    private final List<TypeInfo> bucketTypeInfos;
+    private final int bucketCount;
+    private final int[] bucketColumns;
+    private final int[] partitionColumns;
+    private final List<MethodHandle> hashCodeInvokers;
+
+    public HivePartitionedBucketFunction(
+            BucketingVersion bucketingVersion,
+            int hiveBucketCount,
+            List<HiveType> hiveBucketTypes,
+            List<Type> partitionColumnsTypes,
+            TypeOperators typeOperators,
+            int bucketCount)
+    {
+        this.bucketingVersion = requireNonNull(bucketingVersion, "bucketingVersion is null");
+        this.hiveBucketCount = hiveBucketCount;
+        this.bucketTypeInfos = requireNonNull(hiveBucketTypes, "hiveBucketTypes is null").stream()
+                .map(HiveType::getTypeInfo)
+                .collect(toImmutableList());
+        this.partitionColumns = IntStream.range(hiveBucketTypes.size(), hiveBucketTypes.size() + partitionColumnsTypes.size()).toArray();
+        this.hashCodeInvokers = partitionColumnsTypes.stream()
+                .map(type -> typeOperators.getHashCodeOperator(type, simpleConvention(FAIL_ON_NULL, BLOCK_POSITION)))
+                .collect(toImmutableList());
+        this.bucketColumns = IntStream.range(0, hiveBucketTypes.size()).toArray();
+        this.bucketCount = bucketCount;
+    }
+
+    @Override
+    public int getBucket(Page page, int position)
+    {
+        long partitionHash = 13;
+        for (int i = 0; i < partitionColumns.length; i++) {
+            try {
+                Block partitionColumn = page.getBlock(partitionColumns[i]);
+                partitionHash = 31 * partitionHash + (long) hashCodeInvokers.get(i).invokeExact(partitionColumn, position);
+            }
+            catch (Throwable throwable) {
+                throwIfUnchecked(throwable);
+                throw new RuntimeException(throwable);
+            }
+        }
+
+        long hiveBucket = HiveBucketing.getHiveBucket(bucketingVersion, hiveBucketCount, bucketTypeInfos, page, bucketColumns, position);
+
+        return (int) ((((31 * partitionHash) + hiveBucket) & Long.MAX_VALUE) % bucketCount);
```

Comment:
Why mod `MAX_VALUE`? What is it based on?

Reply:
I looked at HashGenerator::getPartition as reference, it clears the sign bit from the hash value before the modulo with partition count. There is similar logic in hive bucket number calculation as well. It might be because in Java modulo operation retains the sign of the dividend (numerator) and we want to output positive numbers.